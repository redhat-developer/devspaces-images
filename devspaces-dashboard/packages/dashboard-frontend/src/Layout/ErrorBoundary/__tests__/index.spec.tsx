/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { ErrorBoundary, STORAGE_KEY_RELOAD_NUMBER } from '..';
import userEvent from '@testing-library/user-event';

class GoodComponent extends React.Component {
  render() {
    return <span>component</span>;
  }
}
class BadComponent extends React.Component {
  render() {
    throw new Error('Uncaught exception');
    return <span>component</span>;
  }
}
export class NoResourceComponent extends React.Component {
  render() {
    throw new Error('Loading chunk 23 failed.');
    return <span>component</span>;
  }
}

// mute the outputs
console.error = jest.fn();

function wrapComponent(componentToWrap: React.ReactNode) {
  return <ErrorBoundary>{componentToWrap}</ErrorBoundary>;
}

describe('Error boundary', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a correctly working component', () => {
    const goodComponent = <GoodComponent />;
    const errorBoundary = wrapComponent(goodComponent);
    expect(renderer.create(errorBoundary).toJSON()).toMatchSnapshot();
  });

  it('should catch an error thrown inside a component', () => {
    const badComponent = <BadComponent />;
    const errorBoundary = wrapComponent(badComponent);
    expect(renderer.create(errorBoundary).toJSON()).toMatchSnapshot();
  });

  describe('error details', () => {
    it('should show and hide the error stack', () => {
      const errorBoundary = wrapComponent(<BadComponent />);
      render(errorBoundary);

      const showDetailsAction = screen.getByRole('button', { name: 'View stack' });
      userEvent.click(showDetailsAction);

      expect(screen.queryByText('in BadComponent', { exact: false })).toBeTruthy();
      expect(screen.queryByText('in ErrorBoundary', { exact: false })).toBeTruthy();

      userEvent.click(showDetailsAction);

      expect(screen.queryByText('in BadComponent', { exact: false })).toBeFalsy();
      expect(screen.queryByText('in ErrorBoundary', { exact: false })).toBeFalsy();
    });
  });

  describe('when some resources are absent', () => {
    const { reload } = window.location;

    beforeEach(() => {
      delete (window as any).location;
      (window.location as any) = { reload: jest.fn() };

      jest.useFakeTimers();
    });

    afterEach(() => {
      window.location.reload = reload;
      sessionStorage.removeItem(STORAGE_KEY_RELOAD_NUMBER);

      jest.runOnlyPendingTimers();
      jest.useRealTimers();
    });

    it('should inform users that resources are missed', () => {
      const errorBoundary = wrapComponent(<NoResourceComponent />);
      render(errorBoundary);

      expect(
        screen.queryByText('The application has been likely updated on the server.', {
          exact: false,
        }),
      ).toBeTruthy();
    });

    it('should provide the countdown to the next page reload', () => {
      jest.useFakeTimers();

      const errorBoundary = wrapComponent(<NoResourceComponent />);
      render(errorBoundary);

      expect(
        screen.queryByText('Refreshing a page to get newer resources in', { exact: false }),
      ).toBeTruthy();
      expect(screen.queryByText('30 seconds', { exact: false })).toBeTruthy();

      jest.advanceTimersByTime(5000);
      expect(screen.queryByText('25 seconds', { exact: false })).toBeTruthy();
    });

    it('should hide the countdown when stopped', () => {
      const errorBoundary = wrapComponent(<NoResourceComponent />);
      render(errorBoundary);

      const stopCountdownAction = screen.getByRole('button', { name: 'Stop countdown' });
      userEvent.click(stopCountdownAction);

      expect(
        screen.queryByText('Refreshing a page to get newer resources in', { exact: false }),
      ).toBeFalsy();
    });

    it('should not reload page when countdown is stopped', () => {
      const errorBoundary = wrapComponent(<NoResourceComponent />);
      render(errorBoundary);

      const stopCountdownAction = screen.getByRole('button', { name: 'Stop countdown' });
      userEvent.click(stopCountdownAction);

      jest.advanceTimersByTime(35000);
    });

    it('should reload the page when countdown ends', () => {
      const errorBoundary = wrapComponent(<NoResourceComponent />);
      render(errorBoundary);

      jest.advanceTimersByTime(35000);

      expect(window.location.reload).toHaveBeenCalled();
      expect(window.location.reload).toHaveBeenCalledTimes(1);
    });

    it('should reload the page by action', () => {
      const errorBoundary = wrapComponent(<NoResourceComponent />);
      render(errorBoundary);

      const reloadNowAction = screen.getByRole('button', { name: 'Reload now' });
      userEvent.click(reloadNowAction);
      userEvent.click(reloadNowAction);
      userEvent.click(reloadNowAction);

      expect(window.location.reload).toHaveBeenCalled();
      expect(window.location.reload).toHaveBeenCalledTimes(3);
    });

    it('should show additional message after a few reloads', () => {
      const errorBoundary = wrapComponent(<NoResourceComponent />);
      render(errorBoundary);

      const reloadNowAction = screen.getByRole('button', { name: 'Reload now' });

      userEvent.click(reloadNowAction);

      // jest doesn't trigger the `window.onbeforeunload`
      // so it has to be done manually:
      // "refreshing" the page
      window.dispatchEvent(new Event('beforeunload'));
      render(errorBoundary);

      userEvent.click(reloadNowAction);

      // "refreshing" the page
      window.dispatchEvent(new Event('beforeunload'));
      render(errorBoundary);

      expect(
        screen.queryByText(
          'Contact an administrator if refreshing continues after the next load.',
          { exact: true },
        ),
      ).toBeTruthy();
    });

    it('should reset the number of reloads', () => {
      const noResourceComponent = wrapComponent(<NoResourceComponent />);
      render(noResourceComponent);

      const reloadNowAction = screen.getByRole('button', { name: 'Reload now' });

      userEvent.click(reloadNowAction);

      // jest doesn't trigger the `window.onbeforeunload`
      // so it has to be done manually:
      // "refreshing" the page
      window.dispatchEvent(new Event('beforeunload'));
      render(noResourceComponent);

      userEvent.click(reloadNowAction);

      // "refreshing" the page
      window.dispatchEvent(new Event('beforeunload'));
      render(noResourceComponent);

      expect(sessionStorage.getItem(STORAGE_KEY_RELOAD_NUMBER)).toEqual('2');

      const goodComponent = wrapComponent(<GoodComponent />);

      // "refreshing" the page
      window.dispatchEvent(new Event('beforeunload'));
      render(goodComponent);

      // "refreshing" the page
      window.dispatchEvent(new Event('beforeunload'));
      render(goodComponent);

      expect(sessionStorage.getItem(STORAGE_KEY_RELOAD_NUMBER)).toBeNull();
    });
  });
});
