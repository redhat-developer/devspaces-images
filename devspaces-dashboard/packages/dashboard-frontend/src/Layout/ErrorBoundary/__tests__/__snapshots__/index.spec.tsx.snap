// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error boundary should catch an error thrown inside a component 1`] = `
<section
  className="pf-c-page__main-section pf-m-light pf-m-fill"
>
  <div
    aria-label="Danger Alert"
    className="pf-c-alert pf-m-inline pf-m-danger"
    data-ouia-component-id="OUIA-Generated-Alert-danger-1"
    data-ouia-component-type="PF4/Alert"
    data-ouia-safe={true}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="pf-c-alert__icon"
    >
      <svg
        aria-hidden={true}
        aria-labelledby={null}
        fill="currentColor"
        height="1em"
        role="img"
        style={
          Object {
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
        />
      </svg>
    </div>
    <h4
      className="pf-c-alert__title"
    >
      <span
        className="pf-u-screen-reader"
      >
        Danger alert:
      </span>
      Error: Uncaught exception
    </h4>
    <div
      className="pf-c-alert__action-group"
    >
      <button
        aria-disabled={false}
        aria-label={null}
        className="pf-c-button pf-m-link pf-m-inline"
        data-ouia-component-id="OUIA-Generated-Button-link-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        onClick={[Function]}
        role={null}
        type="button"
      >
        View stack
      </button>
    </div>
  </div>
</section>
`;

exports[`Error boundary should render a correctly working component 1`] = `
<span>
  component
</span>
`;
